Ce fixes

fhahn opened the pull request.

fhahn committed:
[ConstraintElimination] Add test case for PR54217.

Adds test case for https://github.com/llvm/llvm-project/issues/54217.

(cherry-picked from 12ffa9c2aa3e)

fhahn committed:
[ConstraintElimination] Only add cond from assume to succs if valid.

Add missing CanAdd check before adding a condition from an assume
to the successor blocks. When adding information from assume to
successor blocks we need to perform the same CanAdd as we do for adding
a condition from a branch.

Fixes https://github.com/llvm/llvm-project/issues/54217

(cherry-picked from c60cdb44f7ec)

fhahn committed:
[ConstraintElimination] Add tests inspired by PR54224.

(cherry-picked from 873e630d6cdb)

fhahn committed:
[ConstraintElimination] Use ZExtValue for unsigned decomposition.

When decomposing constraints for unsigned conditions, we can use
negative values by zero-extending them, as long as they are less than
the maximum constraint value.

Fixes https://github.com/llvm/llvm-project/issues/54224

(cherry-picked from 4bbee17ecb00)

fhahn commented:
@swift-ci please test

fhahn committed:
[ConstraintElimination] Remove over-eager assertion.

After moving the CanAdd check in c60cdb44f7ecb4b02ed and using it for
the assume cases as well, the passed in block may not have  a branch
instruction as terminator. This can trigger the assertion. Given the new
use case, it doesn't add value any longer and can be removed.

Fixes https://github.com/llvm/llvm-project/issues/54281

fhahn commented:
@swift-ci please test

rapidsna commented on llvm/test/Transforms/ConstraintElimination/add-nuw.ll:
```diff
@@ -454,3 +454,127 @@ if.end:                                           ; preds = %entry
+; CHECK-LABEL: @add_nuw_neg_pr54224_i16(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[NEG2:%.*]] = add nuw i16 [[A:%.*]], -305
+; CHECK-NEXT:    [[C_1:%.*]] = icmp ugt i16 0, [[NEG2]]
```
I'm curious why this condition `0 > an unsigned number` is not considered `false` by the compiler.

fhahn replied:
> I'm curious why this condition 0 > an unsigned number is not considered false by the compiler.

Such trivial compares usually are eliminated by other passes before ConstraintElimination runs.

ConstraintElimination could also handle them, if we add a `>= 0` constraint for each unsigned condition we add. But. that's probably better as a separate change, not for `apple:stable/20211026`

rapidsna commented on llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:
```diff
@@ -269,6 +269,19 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT) {
+    auto CanAdd = [&BB, &DT](BasicBlock *Succ) {
+      assert(isa<BranchInst>(BB.getTerminator()));
+      return any_of(successors(&BB),
+                    [Succ](const BasicBlock *S) { return S != Succ; }) &&
```
With this `any_of...`, `CanAdd` returns `false`, if `Succ` is the only successor of `BB`? For example, in the case like below, does it add the assumption for `%cond` to `%succ`?

```
%bb:
  %cond = icmp ugt i16 %a, 0
  assume(%cond)
  br %succ
%succ:
  %cond2 = icmp ugt i16 %a, 0
  return %cond2
```

fhahn replied:
> With this any_of..., CanAdd returns false, if Succ is the only successor of BB? For example, in the case like below, does it add the assumption for %cond to %succ?

That's a good point! We would not add conditions from assumes if there is an unconditional branch as in your example. This can be easily supported and I added another commit to fix it: 12e1762 & 1821ec8

fhahn committed:
[ConstraintElimination] Add additional tests with uncond branches.

(cherry-picked from 2229674cb716)

fhahn committed:
[ConstraintElimination] Support BBs with single successor in CanAdd.

If BB has a single successor, conditions can be added safely.

(cherry-picked from f473d4aa80ce)

fhahn committed:
[ConstraintElimination] Add missing dominance check.

When dealing with an unconditional branch, the condition can only added
if BB properly dominates the successor.

(cherry-picked from 470a975c84f4)

fhahn commented:
@swift-ci please test

rapidsna reviewed (approved):
Thanks @fhahn. LGTM!
